HDDS-10462. Fail Datanode Decommission Early

Tejaskriya opened the pull request:
## What changes were proposed in this pull request?

Many users test out Ozone in small clusters of 15 Datanodes or less. If a 15 DN cluster has some EC 10-4 containers, for example, then it's not possible to decommission more than 1 Datanode, because EC 10-4 requires at least 14 Datanodes. If someone tries to decommission 2 Datanodes, the decommissioning process is designed such that it will keep looping and checking every 30 seconds whether it's possible to take the two DNs offline. It will never fail, even though it's clearly not possible to take the Datanodes offline. 
In this PR, the decommission is failed early if sufficient datanodes are not available based on the maximum replication factor of containers present in the cluster. It returns a corresponding DatanodeAdminError. 
The detailed design doc can be found in the EPIC HDDS-10461

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-10462

## How was this patch tested?

Tested manually in docker cluster. In the following cluster, RATIS-THREE is the maximum replication and it has 5 DNS:
```
bash-4.2$ ozone admin datanode decommission ozone-datanode-4 ozone-datanode-5 ozone-datanode-2
Started decommissioning datanode(s):
ozone-datanode-4
ozone-datanode-5
ozone-datanode-2
Error: AllHosts: Sufficient nodes are not available.
Some nodes could not enter the decommission workflow
``` 
-- will add unit tests and mark it _ready for review_ post that.

siddhantsangwan commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java:
```diff
@@ -368,6 +382,50 @@ public synchronized void startDecommission(DatanodeDetails dn)
+        if (reqNodes > minInService) {
+          minInService = reqNodes;
+        }
+      /* *below code would check the replication type and then get the factor,
+        but as we have a simpler way i.e., getRequiredNodes(), I don't think we need to care about the replication type
+
+      HddsProtos.ReplicationType replicationType = cif.getReplicationType();
+      if (replicationType.equals(HddsProtos.ReplicationType.RATIS)) {
+        maxRatis = cif.getReplicationFactor().getNumber();
+      } else if (replicationType.equals(HddsProtos.ReplicationType.EC) {
+        //cif.getReplicationConfig();
+      }*/
```
Yes, `getRequiredNodes` is what we want.

sodonnel replied:
Yes, getRequiredNodes is the simplest way. Please remove the commented code too (I think you left it there just to ask the question which is perfectly fine).

siddhantsangwan commented on hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/node/TestNodeDecommissionManager.java:
```diff
@@ -56,17 +75,50 @@ public class TestNodeDecommissionManager {
+    pipelineManager.createPipeline(RatisReplicationConfig.getInstance(
+        HddsProtos.ReplicationFactor.THREE));
+    pendingOpsMock = mock(ContainerReplicaPendingOps.class);
+    containerManager = new ContainerManagerImpl(ozConf,
```
I have not checked the test cases yet, but is it necessary to have the actual container manager implementation here? Can we not mock it instead? If we mock it, this test doesn't need to depend on the actual implementation.

Tejaskriya replied:
Thanks for the review, I have used mock for these tests now

siddhantsangwan reviewed (commented):
@Tejaskriya thanks for working on this. We should put this feature behind a feature flag that is CLI driven. The default behaviour would be that decommission fails early when possible, but a user should be able to toggle this off using the decommission command.

sodonnel commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java:
```diff
@@ -368,6 +382,50 @@ public synchronized void startDecommission(DatanodeDetails dn)
+          continue;
+        }
+        int reqNodes = cif.getReplicationConfig().getRequiredNodes();
+        if (reqNodes > minInService) {
```
If you move the line `(nodeManager.getNodeCount(NodeStatus.inServiceHealthy()` to get the total nodes, you don't need to track the minInService. You can just fail early if there are not enough nodes, eg:

```
inService = odeManager.getNodeCount(NodeStatus.inServiceHealthy()) - dns.size();
for (DatanodeDetails dn : dns) {
    ...
    if (cif.getReplicationConfig().getRequiredNodes() > inService) {
        // fail early - no point in checking any further
        return false;
    }
}
```

```

sodonnel commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java:
```diff
@@ -368,6 +382,50 @@ public synchronized void startDecommission(DatanodeDetails dn)
+    for (DatanodeDetails dn : dns) {
+      Set<ContainerID> containers;
+      try {
+        containers = nodeManager.getContainers(dn);
```
It does give me some concern that this large iteration occurs in response to a user command. However, it does not hold any locks beyond blocking further decommission commands, so it is probably OK.

An alternative, is to have Replication Manager track the count of containers per replication_type (and expose them via metrics or the container report) and they we could query RM for the largest number of nodes needed. However the RM data is always stale (runs every 5 minutes), and it cannot tell you if the containers on "this DN" will block. Also RM may not have ran yet after startup.

Tejaskriya replied:
Yes, as you have pointed out, we cannot rely on the RM as using stale data for the fail-early check here as it would be dangerous. The only other way that came up was to iterate through the containers like in this PR

sodonnel commented:
> but a user should be able to toggle this off using the decommission command.

Yea I think we need a --force flag or something to tell it to ignore the check. The reason is that a cluster could have a handful of EC 10-4, but some nodes need to be taken out anyway. The alternative to decommission is to just stop them, and that has to be done one node at a time and wait for replication to complete, so its slow.

If we have the --force flag, then it would gracefully decommission as much as it could, and all the Ratis or smaller EC replication ones. Then it would get stuck and the admin could make a decision to just stop them, knowing the remaining EC containers will not be able to be replicated.

Tejaskriya commented:
@sodonnel  @siddhantsangwan  Thank you for the reviews! I have addressed the comments left. Could you please take a look at the current changes?

siddhantsangwan commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java:
```diff
@@ -368,6 +384,57 @@ public synchronized void startDecommission(DatanodeDetails dn)
+      }
+    }
+
+    for (DatanodeDetails dn : dns) {
```
We need to ignore non in-service DNs in this loop too. Let's add some tests for non in-service and `NodeNotFoundException` DNs.

Tejaskriya replied:
I have added test for checking nodes not-in-service. But in case of NodeNotFound, In the current logic, if we pass the hostname for a non-existent DN then it is caught in `mapHostnamesToDatanodes(nodes, errors)` and the DN is not added to the datanodeDetails list (dns) we iterate through for the checks. Do we need to still consider this scenario?

siddhantsangwan replied:
Yeah, it's not very probable but it's possible that a node goes missing somehow just after the check. We should account for that possibility.

Tejaskriya replied:
I have added a test for this too now using mocks. Could you please review it?

siddhantsangwan reviewed (commented):
I have some comments below. We need tests that cover more scenarios - `force` set to true, DNs that are not in-service or not found in the scm, and any corner cases specified in the design doc that can be unit tested etc. Let's also manually verify the CLI works properly with the new `force` option.

Tejaskriya commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java:
```diff
@@ -368,6 +384,57 @@ public synchronized void startDecommission(DatanodeDetails dn)
+        } catch (ContainerNotFoundException ex) {
+          continue; // ignore the container and continue to next one
+        }
+        if (cif.getState().equals(HddsProtos.LifeCycleState.DELETED) ||
```
I have moved the container related code under a synchronized block. Could you please take a look now?

Tejaskriya commented:
> @Tejaskriya Is this ready for review?

@siddhantsangwan it is ready for review now. I have added few more checks in the same test cases with force flag set to true. The behaviour is as expected. I have also tested it manually in docker set-up, I have updated the results in the PR description under "How was this patch tested"

siddhantsangwan commented on hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/node/TestNodeDecommissionManager.java:
```diff
@@ -369,10 +410,208 @@ public void testNodeDecommissionManagerOnBecomeLeader() throws Exception {
+
+    error = decom.decommissionNodes(Arrays.asList(dns.get(1).getIpAddress(),
+        dns.get(2).getIpAddress(), dns.get(3).getIpAddress(), dns.get(4).getIpAddress()), true);
+    assertTrue(error.size() == 0);
```
Intelllij: `'assertTrue()' can be simplified to 'assertEquals()' `

siddhantsangwan commented on hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/node/TestNodeDecommissionManager.java:
```diff
@@ -369,10 +410,208 @@ public void testNodeDecommissionManagerOnBecomeLeader() throws Exception {
-    scm = HddsTestUtils.getScm(config);
-    return (SCMNodeManager) scm.getScmNodeManager();
+  @Test
+  public void testInsufficientNodeDecommissionThrowsExceptionForRatis() throws
```
The newly added test cases look good. Can we also assert that the DNs have transitioned to the decommissioning state wherever relevant? Existing tests are doing that.

Tejaskriya replied:
Thank you for the review! I have added these checks in the tests. I have addressed the other comments as well. Could you please take a look at this PR again and approve the workflows if everything seems okay?

siddhantsangwan commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java:
```diff
@@ -368,6 +388,63 @@ public synchronized void startDecommission(DatanodeDetails dn)
+        ContainerInfo cif;
+        try {
+          cif = containerManager.getContainer(cid);
+        } catch (ContainerNotFoundException ex) {
```
Let's go with WARN.

Tejaskriya commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java:
```diff
@@ -368,6 +388,63 @@ public synchronized void startDecommission(DatanodeDetails dn)
+  private synchronized boolean checkIfDecommissionPossible(List<DatanodeDetails> dns, List<DatanodeAdminError> errors) {
+    int numDecom = dns.size();
+    List<DatanodeDetails> validDns = dns.stream().collect(Collectors.toList());
+    Collections.copy(validDns, dns);
```
I missed that line during clean up. Thanks for catching it!

siddhantsangwan reviewed (approved):
@Tejaskriya thanks for the update, LGTM. @sodonnel would you like to take another look?

siddhantsangwan commented:
Thanks everyone. Merged to master.
